File: Data Structure/Hash/Medium/LET_no1650_Lowest Common Ancestor of a Binary Tree III.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        memo = set()
        while p:
            memo.add(p)
            p = p.parent
        while q:
            if q in memo:
                return q
            q = q.parent
        return None

File: Data Structure/Hash/Medium/LET_no1650_Lowest Common Ancestor of a Binary Tree III.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* parent;
};
*/

class Solution {
public:
    Node* lowestCommonAncestor(Node* p, Node * q) {
        unordered_set<Node*> memo;
        
        while (p) {
            memo.insert(p);
            p = p->parent;
        }
        
        while (q) {
            if (memo.count(q))
                return q;
            q = q->parent;
        }

        return nullptr;
    }
};